File: py/idkollen_client/models/freja.py
from __future__ import annotations

from typing import Annotated, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from idkollen_client.models.common import ApiErrorCode, Country

FrejaRegistrationLevel = Literal["EXTENDED", "PLUS"]


class _Base(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class FrejaAuthRequest(_Base):
    ssn: Optional[str] = None
    callback_url: Optional[str] = None
    min_registration_level: Optional[FrejaRegistrationLevel] = None
    org_number: Optional[str] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class FrejaBackchannelAuthRequest(_Base):
    ssn: str
    country: Country
    callback_url: Optional[str] = None
    min_registration_level: Optional[FrejaRegistrationLevel] = None
    org_number: Optional[str] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class FrejaSignRequest(_Base):
    text: str
    ssn: Optional[str] = None
    callback_url: Optional[str] = None
    min_registration_level: Optional[FrejaRegistrationLevel] = None
    org_number: Optional[str] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class FrejaBackchannelSignRequest(_Base):
    ssn: str
    country: Country
    text: str
    callback_url: Optional[str] = None
    min_registration_level: Optional[FrejaRegistrationLevel] = None
    org_number: Optional[str] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None


class FrejaPending(_Base):
    status: Literal["PENDING"]
    id: str
    ref_id: Optional[str] = None
    auto_start_token: str
    qr_data: str


class FrejaCompleted(_Base):
    status: Literal["COMPLETED"]
    id: str
    ref_id: Optional[str] = None
    ssn: str
    country: Country
    name: str
    given_name: str
    surname: str
    address: Optional[str] = None
    company_signatory_text: Optional[str] = None


class FrejaFailed(_Base):
    status: Literal["FAILED"]
    id: str
    ref_id: Optional[str] = None
    error: ApiErrorCode


FrejaStatus = Annotated[
    FrejaPending | FrejaCompleted | FrejaFailed,
    Field(discriminator="status"),
]

File: csharp/src/Idkollen.Client/Models/Freja.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Idkollen.Client.Models;

public sealed record FrejaAuthRequest
{
    public string? Ssn { get; init; }
    public string? CallbackUrl { get; init; }
    public string? MinRegistrationLevel { get; init; }
    public string? OrgNumber { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public sealed record FrejaBackchannelAuthRequest(string Ssn, string Country)
{
    public string? CallbackUrl { get; init; }
    public string? MinRegistrationLevel { get; init; }
    public string? OrgNumber { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public sealed record FrejaSignRequest(string Text)
{
    public string? Ssn { get; init; }
    public string? CallbackUrl { get; init; }
    public string? MinRegistrationLevel { get; init; }
    public string? OrgNumber { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public sealed record FrejaBackchannelSignRequest(string Ssn, string Country, string Text)
{
    public string? CallbackUrl { get; init; }
    public string? MinRegistrationLevel { get; init; }
    public string? OrgNumber { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
}

public interface IFrejaStatus { }

public sealed record FrejaPending(
    string Id,
    string? RefId,
    string AutoStartToken,
    string QrData
) : IFrejaStatus;

public sealed record FrejaCompleted(
    string Id,
    string? RefId,
    string Ssn,
    string Country,
    string Name,
    string GivenName,
    string Surname,
    string? Address,
    string? CompanySignatoryText
) : IFrejaStatus;

public sealed record FrejaFailed(string Id, string? RefId, string Error) : IFrejaStatus;

internal sealed class FrejaStatusConverter : JsonConverter<IFrejaStatus>
{
    public override IFrejaStatus Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var status = doc.RootElement.GetProperty("status").GetString();
        var raw = doc.RootElement.GetRawText();
        return status switch
        {
            "PENDING" => JsonSerializer.Deserialize<FrejaPending>(raw, options)!,
            "COMPLETED" => JsonSerializer.Deserialize<FrejaCompleted>(raw, options)!,
            "FAILED" => JsonSerializer.Deserialize<FrejaFailed>(raw, options)!,
            _ => throw new JsonException($"Unknown freja status: {status}"),
        };
    }

    public override void Write(Utf8JsonWriter w, IFrejaStatus v, JsonSerializerOptions o)
        => throw new NotSupportedException();
}
